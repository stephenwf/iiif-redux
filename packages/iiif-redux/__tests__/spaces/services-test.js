import {
  serviceActivate,
  serviceAnnounce,
  serviceDeactivate,
  serviceDisable,
  serviceEnable,
  serviceError,
  reducer,
  DEFAULT_STATE,
} from '../../src/spaces/services';

describe('spaces/services', () => {
  describe('actions', () => {
    it('should create serviceAnnounce action', () => {
      const id = '1234';
      const serviceId = 'http://service/id/';
      const resource = 'http://resource/id/';
      const label = 'test label';
      expect(serviceAnnounce(id, serviceId, resource, label)).toEqual({
        payload: {
          id: '1234',
          label: 'test label',
          resource: 'http://resource/id/',
          serviceId: 'http://service/id/',
        },
        type: 'SERVICE_ANNOUNCE',
      });
    });

    it('should create serviceActivate action', () => {
      const id = '1234';
      const label = 'test label';
      const config = { myCustomProp: 'value' };
      expect(serviceActivate(id, config, { label })).toEqual({
        payload: {
          config: { myCustomProp: 'value' },
          id: '1234',
          label: 'test label',
        },
        type: 'SERVICE_ACTIVATE',
      });
    });

    it('should create serviceError action', () => {
      const id = '1234';
      expect(serviceError(id, 'something went wrong')).toEqual({
        payload: { error: 'something went wrong', id: '1234' },
        type: 'SERVICE_ERROR',
      });
    });

    it('should create serviceDeactivate action', () => {
      const id = '1234';
      expect(serviceDeactivate(id)).toEqual({
        payload: { id: '1234' },
        type: 'SERVICE_DEACTIVATE',
      });
    });

    it('should create serviceEnable action', () => {
      const id = '1234';
      expect(serviceEnable(id)).toEqual({
        payload: { id: '1234' },
        type: 'SERVICE_ENABLE',
      });
    });

    it('should create serviceDisable action', () => {
      const id = '1234';
      expect(serviceDisable(id)).toEqual({
        payload: { id: '1234' },
        type: 'SERVICE_DISABLE',
      });
    });
  });

  describe('reducer', () => {
    it('should handle serviceAnnounce actions', () => {
      const id = '1234';
      const serviceId = 'http://service/id/';
      const resource = 'http://resource/id/';
      const label = 'test label';

      expect(
        [serviceAnnounce(id, serviceId, resource, label)].reduce(
          reducer,
          DEFAULT_STATE
        )
      ).toMatchSnapshot();
    });

    it('should handle serviceAnnounce and then serviceActivate actions', () => {
      const id = '1234';
      const serviceId = 'http://service/id/';
      const resource = 'http://resource/id/';
      const label = 'test label';
      const config = { myCustomProp: 'some value' };

      expect(
        [
          serviceAnnounce(id, serviceId, resource, label),
          serviceActivate(id, config),
        ].reduce(reducer, DEFAULT_STATE)
      ).toMatchSnapshot();
    });

    it('should be able to override label when activating service', () => {
      const id = '1234';
      const serviceId = 'http://service/id/';
      const resource = 'http://resource/id/';
      const label = 'test label';
      const config = { myCustomProp: 'some value' };

      expect(
        [
          serviceAnnounce(id, serviceId, resource, label),
          serviceActivate(id, config, { label: 'test override label' }),
        ].reduce(reducer, DEFAULT_STATE)
      ).toMatchSnapshot();
    });

    it('should handle serviceAnnounce, serviceActivate and then serviceEnable actions', () => {
      const id = '1234';
      const serviceId = 'http://service/id/';
      const resource = 'http://resource/id/';
      const label = 'test label';
      const config = { myCustomProp: 'some value' };

      expect(
        [
          serviceAnnounce(id, serviceId, resource, label),
          serviceActivate(id, config),
          serviceEnable(id),
        ].reduce(reducer, DEFAULT_STATE)
      ).toMatchSnapshot();
    });

    it('should handle serviceAnnounce, serviceActivate, serviceEnable and then serviceDisable actions', () => {
      const id = '1234';
      const serviceId = 'http://service/id/';
      const resource = 'http://resource/id/';
      const label = 'test label';
      const config = { myCustomProp: 'some value' };

      expect(
        [
          serviceAnnounce(id, serviceId, resource, label),
          serviceActivate(id, config),
          serviceEnable(id),
          serviceDisable(id),
        ].reduce(reducer, DEFAULT_STATE)
      ).toMatchSnapshot();
    });

    it('should deactivate and disable service after error', () => {
      const id = '1234';
      const serviceId = 'http://service/id/';
      const resource = 'http://resource/id/';
      const label = 'test label';
      const config = { myCustomProp: 'some value' };

      expect(
        [
          serviceAnnounce(id, serviceId, resource, label),
          serviceActivate(id, config),
          serviceEnable(id),
          serviceError(id, 'Something went wrong'),
        ].reduce(reducer, DEFAULT_STATE)
      ).toMatchSnapshot();
    });

    it('should be able to deactivate service', () => {
      const id = '1234';
      const serviceId = 'http://service/id/';
      const resource = 'http://resource/id/';
      const label = 'test label';
      const config = { myCustomProp: 'some value' };

      expect(
        [
          serviceAnnounce(id, serviceId, resource, label),
          serviceActivate(id, config),
          serviceEnable(id),
          serviceDeactivate(id),
        ].reduce(reducer, DEFAULT_STATE)
      ).toMatchSnapshot();
    });

    it('should be able to handle multiple resources', () => {
      expect(
        [
          serviceAnnounce('1', 'service-1', 'resource-1', 'label 1'),
          serviceAnnounce('2', 'service-2', 'resource-2', 'label 2'),
          serviceAnnounce('3', 'service-3', 'resource-3', 'label 3'),
        ].reduce(reducer, DEFAULT_STATE)
      ).toMatchSnapshot();
    });

    it('should be able to handle multiple services on a resource', () => {
      expect(
        [
          serviceAnnounce('1', 'service-1', 'resource-1', 'label 1'),
          serviceAnnounce('2', 'service-2', 'resource-1', 'label 2'),
          serviceAnnounce('3', 'service-3', 'resource-1', 'label 3'),
        ].reduce(reducer, DEFAULT_STATE)
      ).toMatchSnapshot();
    });
  });
});
