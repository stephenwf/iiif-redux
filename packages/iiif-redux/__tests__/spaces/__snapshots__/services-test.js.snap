// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`spaces/services reducer should be able to deactivate service 1`] = `
Object {
  "list": Object {
    "1234": Object {
      "config": Object {
        "myCustomProp": "some value",
      },
      "id": "1234",
      "isActive": false,
      "isEnabled": false,
      "label": "test label",
      "resource": "http://resource/id/",
      "serviceId": "http://service/id/",
    },
  },
  "resourceIndex": Object {
    "http://resource/id/": Array [
      "1234",
    ],
  },
}
`;

exports[`spaces/services reducer should be able to handle multiple resources 1`] = `
Object {
  "list": Object {
    "1": Object {
      "id": "1",
      "isActive": false,
      "isEnabled": false,
      "label": "label 1",
      "resource": "resource-1",
      "serviceId": "service-1",
    },
    "2": Object {
      "id": "2",
      "isActive": false,
      "isEnabled": false,
      "label": "label 2",
      "resource": "resource-2",
      "serviceId": "service-2",
    },
    "3": Object {
      "id": "3",
      "isActive": false,
      "isEnabled": false,
      "label": "label 3",
      "resource": "resource-3",
      "serviceId": "service-3",
    },
  },
  "resourceIndex": Object {
    "resource-1": Array [
      "1",
    ],
    "resource-2": Array [
      "2",
    ],
    "resource-3": Array [
      "3",
    ],
  },
}
`;

exports[`spaces/services reducer should be able to handle multiple services on a resource 1`] = `
Object {
  "list": Object {
    "1": Object {
      "id": "1",
      "isActive": false,
      "isEnabled": false,
      "label": "label 1",
      "resource": "resource-1",
      "serviceId": "service-1",
    },
    "2": Object {
      "id": "2",
      "isActive": false,
      "isEnabled": false,
      "label": "label 2",
      "resource": "resource-1",
      "serviceId": "service-2",
    },
    "3": Object {
      "id": "3",
      "isActive": false,
      "isEnabled": false,
      "label": "label 3",
      "resource": "resource-1",
      "serviceId": "service-3",
    },
  },
  "resourceIndex": Object {
    "resource-1": Array [
      "1",
      "2",
      "3",
    ],
  },
}
`;

exports[`spaces/services reducer should be able to override label when activating service 1`] = `
Object {
  "list": Object {
    "1234": Object {
      "config": Object {
        "myCustomProp": "some value",
      },
      "id": "1234",
      "isActive": true,
      "isEnabled": false,
      "label": "test override label",
      "resource": "http://resource/id/",
      "serviceId": "http://service/id/",
    },
  },
  "resourceIndex": Object {
    "http://resource/id/": Array [
      "1234",
    ],
  },
}
`;

exports[`spaces/services reducer should deactivate and disable service after error 1`] = `
Object {
  "list": Object {
    "1234": Object {
      "config": Object {
        "myCustomProp": "some value",
      },
      "error": "Something went wrong",
      "id": "1234",
      "isActive": false,
      "isEnabled": false,
      "label": "test label",
      "resource": "http://resource/id/",
      "serviceId": "http://service/id/",
    },
  },
  "resourceIndex": Object {
    "http://resource/id/": Array [
      "1234",
    ],
  },
}
`;

exports[`spaces/services reducer should handle serviceAnnounce actions 1`] = `
Object {
  "list": Object {
    "1234": Object {
      "id": "1234",
      "isActive": false,
      "isEnabled": false,
      "label": "test label",
      "resource": "http://resource/id/",
      "serviceId": "http://service/id/",
    },
  },
  "resourceIndex": Object {
    "http://resource/id/": Array [
      "1234",
    ],
  },
}
`;

exports[`spaces/services reducer should handle serviceAnnounce and then serviceActivate actions 1`] = `
Object {
  "list": Object {
    "1234": Object {
      "config": Object {
        "myCustomProp": "some value",
      },
      "id": "1234",
      "isActive": true,
      "isEnabled": false,
      "label": "test label",
      "resource": "http://resource/id/",
      "serviceId": "http://service/id/",
    },
  },
  "resourceIndex": Object {
    "http://resource/id/": Array [
      "1234",
    ],
  },
}
`;

exports[`spaces/services reducer should handle serviceAnnounce, serviceActivate and then serviceEnable actions 1`] = `
Object {
  "list": Object {
    "1234": Object {
      "config": Object {
        "myCustomProp": "some value",
      },
      "id": "1234",
      "isActive": true,
      "isEnabled": true,
      "label": "test label",
      "resource": "http://resource/id/",
      "serviceId": "http://service/id/",
    },
  },
  "resourceIndex": Object {
    "http://resource/id/": Array [
      "1234",
    ],
  },
}
`;

exports[`spaces/services reducer should handle serviceAnnounce, serviceActivate, serviceEnable and then serviceDisable actions 1`] = `
Object {
  "list": Object {
    "1234": Object {
      "config": Object {
        "myCustomProp": "some value",
      },
      "id": "1234",
      "isActive": true,
      "isEnabled": false,
      "label": "test label",
      "resource": "http://resource/id/",
      "serviceId": "http://service/id/",
    },
  },
  "resourceIndex": Object {
    "http://resource/id/": Array [
      "1234",
    ],
  },
}
`;
